File: SRM144/PowerOutage.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Text;

public class PowerOutage {
	public int estimateTimeOut(int[] fromJunction, int[] toJunction, int[] ductLength) {
		return 0;
	}

// CUT begin
	static bool DoTest(int[] fromJunction, int[] toJunction, int[] ductLength, int __expected) {
		DateTime startTime = DateTime.Now;
		Exception exception = null;
		PowerOutage instance = new PowerOutage();
		int __result = 0;
		try {
			__result = instance.estimateTimeOut(fromJunction, toJunction, ductLength);
		}
		catch (Exception e) { exception = e; }
		TimeSpan __elapsed = new TimeSpan(DateTime.Now.Ticks - startTime.Ticks);

		if (exception != null) {
			Console.Error.WriteLine("RUNTIME ERROR!");
			Console.Error.WriteLine(exception);
			return false;
		}
		else if (__result == __expected) {
			Console.Error.WriteLine("PASSED! " + string.Format("({0:0.00} seconds)", __elapsed.TotalSeconds));
			return true;
		}
		else {
			Console.Error.WriteLine("FAILED! " + string.Format("({0:0.00} seconds)", __elapsed.TotalSeconds));
			Console.Error.WriteLine("           Expected: " + __expected);
			Console.Error.WriteLine("           Received: " + __result);
			return false;
		}
	}

	public static void Main(string[] args){
		HashSet<int> cases = new HashSet<int>();
        for (int i = 0; i < args.Length; ++i) cases.Add(int.Parse(args[i]));

		Console.Error.WriteLine("PowerOutage (1100 Points)");
		Console.Error.WriteLine();

    	int nCases = 0, nPassed = 0;
    	using (var reader = File.OpenText("..\\..\\..\\..\\SRM144\\PowerOutage.sample")) {
            while (true) {
                string line = reader.ReadLine();
                if (line == null || !line.StartsWith("--"))
                    break;

                int[] fromJunction = new int[int.Parse(reader.ReadLine())];
                for (int i = 0; i < fromJunction.Length; ++i)
                    fromJunction[i] = (int) Convert.ChangeType(reader.ReadLine(), typeof(int));
                int[] toJunction = new int[int.Parse(reader.ReadLine())];
                for (int i = 0; i < toJunction.Length; ++i)
                    toJunction[i] = (int) Convert.ChangeType(reader.ReadLine(), typeof(int));
                int[] ductLength = new int[int.Parse(reader.ReadLine())];
                for (int i = 0; i < ductLength.Length; ++i)
                    ductLength[i] = (int) Convert.ChangeType(reader.ReadLine(), typeof(int));
                reader.ReadLine();
                int __answer = (int) Convert.ChangeType(reader.ReadLine(), typeof(int));

                nCases++;
                if (cases.Count > 0 && !cases.Contains(nCases - 1))
                	continue;
                Console.Error.Write(string.Format("  Testcase #{0} ... ", nCases - 1));
                if (DoTest(fromJunction, toJunction, ductLength, __answer))
                    nPassed++;
            }
    	}

        if (cases.Count > 0) nCases = cases.Count;
		Console.Error.WriteLine();
		Console.Error.WriteLine(string.Format("Passed : {0}/{1} cases", nPassed, nCases));

		DateTime Jan1st1970 = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
		int T = (int)((DateTime.UtcNow - Jan1st1970).TotalSeconds - 1455864129);
		double PT = T / 60.0, TT = 75.0;
		Console.Error.WriteLine(string.Format("Time   : {0} minutes {1} secs", T / 60, T % 60));
		Console.Error.WriteLine(string.Format("Score  : {0:0.00} points", 1100 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT))));
	}
// CUT end
}

File: SRM144/PowerOutage.cpp
#include <cstdio>
#include <cmath>
#include <cstring>
#include <ctime>
#include <iostream>
#include <algorithm>
#include <set>
#include <vector>
#include <sstream>
#include <typeinfo>
#include <fstream>
#include <numeric>

using namespace std;

const static int MAX_ELEMENT = 50;
const static int INFTY = 9999999;

class PowerOutage {
	public:
	int estimateTimeOut(vector<int> fromJunction, vector<int> toJunction, vector<int> ductLength) {
		vector<vector<int> > J(MAX_ELEMENT, vector<int>(MAX_ELEMENT, INFTY));
		int elements = fromJunction.size();
		for (int i = 0; i < elements; i++)
		{
			J[fromJunction[i]][toJunction[i]] = J[toJunction[i]][fromJunction[i]] = ductLength[i];
		}

		int farthestDis = dijkstra(J);
		int total = accumulate(ductLength.begin(), ductLength.end(), 0);
		int minMinutes = 2 * total - farthestDis;
        return minMinutes;
    }

	private:
	int dijkstra(const vector<vector<int> >& J){
		vector<int> D(MAX_ELEMENT);
		for (int i = 1; i < MAX_ELEMENT; i++)
		{
			D[i] = J[0][i];
		}

		vector<bool> visited(MAX_ELEMENT, false);
		while (true){
			int minDis = INFTY;
			int minIdx = -1;
			for (int j = 1; j < MAX_ELEMENT; j++){
				if (!visited[j] && D[j] < minDis){
					minDis = D[j];
					minIdx = j;
				}
			}
			if (minIdx == -1)
				break;
			visited[minIdx] = true;

			for (int j = 1; j < MAX_ELEMENT; j++){
				D[j] = min(D[j], minDis + J[minIdx][j]);
			}
		}

		int maxDis = 0;
		for (int i = 0; i < MAX_ELEMENT; i++){
			if (D[i] != INFTY && D[i] > maxDis)
				maxDis = D[i];
		}
		return maxDis;
	}
};

// CUT begin
ifstream data("..\\..\\SRM144\\PowerOutage.sample");

string next_line() {
    string s;
    getline(data, s);
    return s;
}

template <typename T> void from_stream(T &t) {
    stringstream ss(next_line());
    ss >> t;
}

void from_stream(string &s) {
    s = next_line();
}

template <typename T> void from_stream(vector<T> &ts) {
    int len;
    from_stream(len);
    ts.clear();
    for (int i = 0; i < len; ++i) {
        T t;
        from_stream(t);
        ts.push_back(t);
    }
}

template <typename T>
string to_string(T t) {
    stringstream s;
    s << t;
    return s.str();
}

string to_string(string t) {
    return "\"" + t + "\"";
}

bool do_test(vector<int> fromJunction, vector<int> toJunction, vector<int> ductLength, int __expected) {
    time_t startClock = clock();
    PowerOutage *instance = new PowerOutage();
    int __result = instance->estimateTimeOut(fromJunction, toJunction, ductLength);
    double elapsed = (double)(clock() - startClock) / CLOCKS_PER_SEC;
    delete instance;

    if (__result == __expected) {
        cout << "PASSED!" << " (" << elapsed << " seconds)" << endl;
        return true;
    }
    else {
        cout << "FAILED!" << " (" << elapsed << " seconds)" << endl;
        cout << "           Expected: " << to_string(__expected) << endl;
        cout << "           Received: " << to_string(__result) << endl;
        return false;
    }
}

int run_test(bool mainProcess, const set<int> &case_set, const string command) {
    int cases = 0, passed = 0;
    while (true) {
        if (next_line().find("--") != 0)
            break;
        vector<int> fromJunction;
        from_stream(fromJunction);
        vector<int> toJunction;
        from_stream(toJunction);
        vector<int> ductLength;
        from_stream(ductLength);
        next_line();
        int __answer;
        from_stream(__answer);

        cases++;
        if (case_set.size() > 0 && case_set.find(cases - 1) == case_set.end())
            continue;

        cout << "  Testcase #" << cases - 1 << " ... ";
        if ( do_test(fromJunction, toJunction, ductLength, __answer)) {
            passed++;
        }
    }
    if (mainProcess) {
        cout << endl << "Passed : " << passed << "/" << cases << " cases" << endl;
        int T = time(NULL) - 1455864115;
        double PT = T / 60.0, TT = 75.0;
        cout << "Time   : " << T / 60 << " minutes " << T % 60 << " secs" << endl;
        cout << "Score  : " << 1100 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT)) << " points" << endl;
    }
    return 0;
}

int main(int argc, char *argv[]) {
    cout.setf(ios::fixed, ios::floatfield);
    cout.precision(2);
    set<int> cases;
    bool mainProcess = true;
    for (int i = 1; i < argc; ++i) {
        if ( string(argv[i]) == "-") {
            mainProcess = false;
        } else {
            cases.insert(atoi(argv[i]));
        }
    }
    if (mainProcess) {
        cout << "PowerOutage (1100 Points)" << endl << endl;
    }
    return run_test(mainProcess, cases, argv[0]);
}
// CUT end
